### PR TITLE
correct behaviour for assert_array_less with inf/nan

### LINKED ISSUE #2418: assert_array_less behavior with inf/nan (Trac #1825)
_Original ticket http://projects.scipy.org/numpy/ticket/1825 on 2011-05-08 by @rgommers, assigned to unknown._

This passed with 1.5.1 but fails with 1.6.0 due to Inf fixes to assert_array_compare for functions testing (near-)equality.

```
np.testing.assert_array_less(2, np.inf)
np.testing.assert_array_less(np.array([ 0.911,  1.065,  1.325,  1.587]), np.inf)
```

These functions pass butit's not clear that they should:

```
assert_array_less([1, np.inf], [2, np.inf])
assert_array_less([1, np.nan], [2, np.nan])
```

### PR DESCRIPTION
**Pull request by AnishShah**
Fixes #2418

### CODE REVIEW

**Comment by AnishShah** [comment]
@rgommers Can you please review this PR?

File: numpy/testing/utils.py [hunk 1]
```diff
@@ -689,20 +689,13 @@ def chk_same_position(x_id, y_id, hasval='nan'):
...
             # Validate that the special values are in the same place
             if any(x_isnan) or any(y_isnan):
                 chk_same_position(x_isnan, y_isnan, hasval='nan')
-            if any(x_isinf) or any(y_isinf):
-                # Check +inf and -inf separately, since they are different
-                chk_same_position(x == +inf, y == +inf, hasval='+inf')
-                chk_same_position(x == -inf, y == -inf, hasval='-inf')
 
             # Combine all the special values
             x_id, y_id = x_isnan, y_isnan
```
Comment (madphysicist) [inline numpy/testing/utils.py]: Why not just `x_isnan | x_isinf, y_isnan | y_isinf`? I would find that just a little less confusing than trying to figure out what the next two lines do.

File: numpy/testing/utils.py [hunk 2]
```diff
@@ -689,20 +689,13 @@ def chk_same_position(x_id, y_id, hasval='nan'):
 
         if isnumber(x) and isnumber(y):
             x_isnan, y_isnan = isnan(x), isnan(y)
-            x_isinf, y_isinf = isinf(x), isinf(y)
```
Comment (madphysicist) [inline numpy/testing/utils.py]: What was the motivation to remove the checks for +/-infinity?

Comment (AnishShah) [inline numpy/testing/utils.py]: Previous logic for `assert_array_compare` was if there was `inf` in `x` or `y`, it should be at the same location (array id). It just checks the location of `inf`, it does not do any sort of comparison, which is wrong.

Comment (madphysicist) [inline numpy/testing/utils.py]: So just for my edification, you are relying on the following being true then:

    +inf == +inf
    -inf == -inf
    -inf < +inf
    -inf < some_finite_number
    some_finite_number < +inf

Comment (AnishShah) [inline numpy/testing/utils.py]: Yes

Comment (madphysicist) [inline numpy/testing/utils.py]: Thanks for the explanation.

Comment (AnishShah) [inline numpy/testing/utils.py]: Welcome :)

**Comment by AnishShah** [comment]
any review for this PR?

**Comment by charris** [comment]
This is not correct, look at the compare function passed in `assert_array_almost_equal` and see that in the following `a` and `b` are almost equal.
```
In [1]: from numpy.testing import assert_array_almost_equal

In [2]: a = array([np.inf, -np.inf])

In [3]: b = array([-np.inf, np.inf])

In [4]: assert_array_almost_equal(a, b)
```

**Pull request closed without merging**